cpp/src/streaming/core/memory_reserve_or_wait.cpp

```diff
@@ -0,0 +1,258 @@
+    if (ret.size() < size) {
+        co_return ctx_->br()->reserve(mem_type_, size, /* allow_overbooking = */ true);
+    }
+    co_return {std::move(ret), size};
+}
+
+std::size_t MemoryReserveOrWait::size() const {
+    std::lock_guard lock(mutex_);
+    return reservation_requests_.size();
+}
+
+std::size_t MemoryReserveOrWait::periodic_memory_check_counter() const {
+    return periodic_memory_check_counter_.load(std::memory_order_acquire);
+}
+
+coro::task<void> MemoryReserveOrWait::periodic_memory_check() {
+    BufferResource* br = ctx_->br();
+
+    // Helper that returns available memory, clamped so negative values become zero.
+    auto memory_available = [f = br->memory_available(mem_type_)]() -> std::size_t {
+        std::int64_t const ret = f();
+        return static_cast<std::size_t>(std::max(ret, std::int64_t{0}));
+    };
+
+    // Helper that returns the subrange of reservation requests with size <= max_size.
+    auto eligible_requests = [this](std::size_t max_size)
+        -> std::ranges::subrange<std::set<Request>::const_iterator> {
+        // Since `reservation_requests_` is sorted by ascending size,
+        // upper_bound finds the first element with size > max_size.
+        auto last = std::ranges::upper_bound(
+            reservation_requests_, max_size, std::less<>{}, &Request::size
+        );
+        // The range [begin, last) contains all requests with size <= max_size.
+        return {reservation_requests_.begin(), last};
+    };
+
+    // Helper that pushes a memory reservation into a request's queue **without**
+    // waiting on the coroutine.
+    auto push_into_queue = [this](Request& request, MemoryReservation res) -> void {
+        auto err = ctx_->executor()->spawn_detached(
+            [](coro::queue<MemoryReservation>& queue, MemoryReservation res) -> Node {
+                RAPIDSMPF_EXPECTS(
+                    co_await queue.push(std::move(res))
+                        == coro::queue_produce_result::produced,
+                    "could not push memory reservation"
+                );
+            }(request.queue, std::move(res))
+        );
+        RAPIDSMPF_EXPECTS(err, "cannot spawn push-into-queue task");
+    };
+
+    while (true) {
+        auto last_reservation_success = Clock::now();
+        while (true) {
+            // Exit if no more pending requests remain.
+            {
+                std::unique_lock lock(mutex_);
+                if (reservation_requests_.empty()) {
+                    co_return;
+                }
+            }
+            periodic_memory_check_counter_.fetch_add(1, std::memory_order_acq_rel);
+            co_await ctx_->executor()->yield();
+            if (Clock::now() - last_reservation_success > timeout_) {
+                // This is the only way out of the while-loop that doesn't shutdown
+                // the periodic memory check.
+                break;
+            }
```

Comment:
I'm missing something here. Shouldnt `timeout_` be evaluated for each request? `periodic_memory_check` is a long-running task per `MemoryReserveOrWait` instance. So, when the timeout expires, we pick the smallest request with the future_release_potential, and try to reserve that. BUT this could actually be the _last_  request that was added. And if the reservation fails, we inform it that it has failed. But in reality, it could have been attempted again later. 

What I'm trying to say is that, I feel like the Request should have a timestamp that tells us when it was created, and based on time-now, any expired requests should also be released.